app/models/semester_day.rb

```diff
@@ -0,0 +1,64 @@
+require 'day'
+require 'faculty_semester'
+
+class SemesterDay
+
+  attr_reader :date, :semester
+
+  def initialize(semester, date)
+    @semester = semester.freeze
+    @date = date.freeze
+  end
+
+  def self.resolve(date, faculty_id)
+    if (semester = FacultySemester.find_by_date(date, faculty_id))
+      new semester, date
+    end
+  end
+
+  def wday
+    return date.wday unless period.first_day_override
+
+    @wday ||= begin
+      offset = Day.to_numeric(period.first_day_override) - period.starts_at.wday
+      (date.wday + offset) % 7
+    end
+  end
+
+  def wday_name
+    Day.from_numeric(wday)
+  end
+
+  def week_parity
+    return if period.type != :teaching
+
+    @week_parity ||= begin
+      first_parity = period.first_week_parity == :even ? 0 : 1
+      parity_name(period.starts_at.cweek + date.cweek + first_parity)
+    end
+  end
+
+  def period
+    @period ||= @semester.semester_periods.find do |p|
+      date >= p.starts_at  && date <= p.ends_at
+    end
+  end
+
+  def week_num
+    @week_num ||= begin
+      periods = @semester.semester_periods.find_all do |p|
```

Comment:
Did you consider possibility that `semester_periods` may not always be ordered by date in database?